include/vcpkg/base/cofffilereader.h

```diff
@@ -40,7 +230,61 @@ namespace vcpkg
 
     MachineType to_machine_type(const uint16_t value);
 
-    MachineType read_dll_machine_type(const ReadFilePointer& fs);
+    enum class PEType
+    {
+        Unset,
+        PE32,
+        PE32Plus
+    };
+
+    enum class LoadConfigType
+    {
+        UnsetOrOld,
+        PE32,
+        PE32Plus
+    };
+
+    struct DllMetadata
+    {
+        CoffFileHeader coff_header;
+        PEType pe_type;
+        CommonPEOptionalHeaders common_optional_headers;
+        union
+        {
+            UniquePEOptionalHeaders pe_headers;
+            UniquePEPlusOptionalHeaders pe_plus_headers;
+        };
+
+        LoadConfigType load_config_type;
+        union
+        {
+            ImageLoadConfigDirectory32 image_config_directory32;
+            ImageLoadConfigDirectory64 image_config_directory64;
+        };
+
+        std::vector<ImageDataDirectory> data_directories;
+        std::vector<SectionTableHeader> section_headers;
+
+        MachineType get_machine_type() const noexcept;
+        bool is_arm64_ec() const noexcept;
+    };
+
+    struct ArchiveMemberHeader
+    {
+        char name[16];
+        char date[12];
+        char user_id[6];
+        char group_id[6];
+        char mode[8];
+        char size[10];
+        char end_of_header[2];
+
+        void check_end_of_header() const;
+        uint64_t decoded_size() const;
+    };
 
-    std::vector<MachineType> read_lib_machine_types(const ReadFilePointer& fs);
+    ExpectedL<DllMetadata> try_read_dll_metadata(PositionedReadPointerU32& f);
```

Comment:
Perhaps this should accept an interface rather than only `PositionedReadPointerU32` accept no substitutes?